Docs for State Refactor / State Management (#1206)
ceramicnetwork/js-ceramic: docs-src/adr/1-state-refactor.adr.md

# State Refactor

* Status: accepted
* Deciders: Sergey Ukustov, Spencer Brody, Joel Thorstensson

> * Date: 2020-03-31

Reviewer: ```suggestion
* Date: 2021-03-31
```

Reviewer: We're finally free of 2020!  No going back!  😂

* Components communicate asynchronously via JS events. While it is standard in Node.js environment, we can not use the same mechanism on browser without external dependencies. We would like to have most of our code isomorphic.
* One could not easily compose event-driven components. Code gets less comprehensible with every component. We want to have rxjs as a solution for the composability problem.

## Considered Options

In any case, we maintain a LRU cache of documents to achieve a goal of constant memory consumption. This constrains us in a sense that a document in the cache is ephemeral. It can be evicted at any time. The big decision is what concurrency model to adopt for document operations so that it plays nicely with cache eviction.

Document operations can be invoked from few places. One is external API commands: HTTP endpoints of a Ceramic node. Another is JS API: calling in-process Ceramic node. The third one is IPFS pubsub. Also, there is anchoring mechanism, that is a continuous polling of the anchor service. These could be started concurrently.

> We want to maintain document state consistency in the presence of concurrent operations. Under no circumstances two operations are allowed to change the document at the same time.

Reviewer: ```suggestion
We want to maintain document state consistency in the presence of concurrent operations. Under no circumstances two operations are allowed to change the same document at the same time.
```

* Option 1: On cache eviction wait for the operations on a document to finish.

>   With the LRU semantics, it means calling asynchronous function on an evicted document (`await document.close`). Since we do not want to end up with having two documents running in memory at the same time, cache eviction effectively blocks further documents to be processed until the evicted one is finished.

Reviewer: ```suggestion
  With the LRU semantics, it means calling asynchronous function on an evicted document (`await document.close`). Since we do not want to end up with having more documents running in memory at the same time than the configured cache size, cache eviction effectively blocks further documents to be processed until the evicted one is finished.
```

Author: @stbrody I tried to emphasise different scenario - two instances with the same docId - but this one is important too.

So, now we have two explicit reasons:
> With the LRU semantics, it means calling asynchronous function on an evicted document (`await document.close`). Since we do not want to end up with having two instances of the same document running in memory at the same time (potential conflict) or having more documents in memory than the configured cache size, cache eviction effectively blocks further documents to be processed until the evicted one is finished.

  Pro:

  - simple concurrency model

  Contra:

  - document processing is blocked till the oldest document finishes

* Option 2: Separate operation on a document from its in-memory lifecycle.

  The separation is achieved by introducing an execution queue. The execution queue ensures operations on the same docId are executed sequentially. Before an operation is started, the queue loads a document in memory. When a document is loaded, the oldest document gets evicted. This way we maintain a limited number of documents running in memory while not caring about document processing. We further limit this vector of memory growth by queuing the processing tasks.

  Pro:

  - more controllable concurrency model,
  - non-blocking cache eviction and new document processing,

  Contra:

  - more parameters to control concurrency,
  - more complex flow.

## Decision Outcome

We chosen option 2. It allows us to limit memory growth by constant factor, and enables more tight control over different vectors of memory growth.

### Positive Consequences

> - We have configurable number of documents running in memory - ceiling on memory growth when processing load is low.

Reviewer: ```suggestion
- We have configurable number of documents running in memory (`CeramicConfig.docCacheLimit`) - ceiling on memory growth when processing load is low.
```

> - We have configurable number of concurrent document-processing tasks - ceiling on memory growth when processing load is high.

Reviewer: ```suggestion
- We have configurable number of concurrent document-processing tasks (`CeramicConfig.concurrentRequestsLimit`) - ceiling on memory growth when processing load is high.
```
